kvdb/cache.go

```diff
@@ -0,0 +1,930 @@
+	"container/list"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+	"github.com/google/btree"
+)
+
+const (
+	treeDeg = 3
+)
+
+type cacheBucket struct {
+	seq  *uint64
+	tree *btree.BTree
+}
+
+func newCacheBucket() *cacheBucket {
+	return &cacheBucket{
+		tree: btree.New(treeDeg),
+	}
+}
+
+func (c *cacheBucket) get(key []byte) *cachedItem {
+	if key == nil {
+		return nil
+	}
+
+	keyItem := &cachedItem{
+		key: string(key),
+	}
+
+	if valItem := c.tree.Get(keyItem); valItem != nil {
+		return valItem.(*cachedItem)
+	}
+
+	return nil
+}
+
+func (c *cacheBucket) put(item *cachedItem) {
+	c.tree.ReplaceOrInsert(item)
+}
+
+func (c *cacheBucket) del(item *cachedItem) {
+	c.tree.Delete(item)
+}
+
+type cachedItem struct {
+	key    string
+	value  string
+	bucket *cacheBucket
+}
+
+func (c *cachedItem) Less(than btree.Item) bool {
+	return c.key < than.(*cachedItem).key
+}
+
+// Enforce that Cache implements the ExtendedBackend interface.
+var _ walletdb.DB = (*Cache)(nil)
+
+type pendingChange interface {
+	// Reverts the pending change.
+	Revert()
+}
+
+type pendingAdd struct {
+	parent   *cachedItem
+	newChild *cachedItem
+}
+
+func (p *pendingAdd) Revert() {
+	p.parent.bucket.tree.Delete(p.newChild)
+}
+
+type pendingUpdate struct {
+	parent   *cachedItem
+	oldValue string
+}
+
+func (p *pendingUpdate) Revert() {
+	p.parent.value = p.oldValue
+}
+
+type pendingDelete struct {
+	parent   *cachedItem
+	oldChild *cachedItem
+}
+
+func (p *pendingDelete) Revert() {
+	p.parent.bucket.tree.ReplaceOrInsert(p.oldChild)
+}
+
+// Cache is a simple write through cache implementing the kvdb.Backend
+// interface. It's capable of recursively caching top-level buckets speeding up
+// reads by reducing roundtrips to the actual DB backend while remaining
+// consistend with DB state as long as the cache is used when mutating those
+// buckets.
+type Cache struct {
+	mx       sync.RWMutex
+	currRwTx *cacheReadWriteTx
+	root     *cachedItem
+	pending  *list.List
+	backend  Backend
+}
+
+// NewCache constructs a new cache.
+func NewCache(backend Backend) *Cache {
+	return &Cache{
+		root: &cachedItem{
+			bucket: newCacheBucket(),
+		},
+		pending: list.New(),
+		backend: backend,
+	}
+}
+
+func (c *Cache) pendingAdd(parent *cachedItem, newChild *cachedItem) {
+	c.pending.PushBack(&pendingAdd{parent, newChild})
+}
+
+func (c *Cache) pendingUpdate(parent *cachedItem, oldValue string) {
+	c.pending.PushBack(&pendingUpdate{parent, oldValue})
+}
+
+func (c *Cache) pendingDelete(parent *cachedItem, oldChild *cachedItem) {
+	c.pending.PushBack(&pendingDelete{parent, oldChild})
+}
+
+type traversalHelper struct {
+	bucket walletdb.ReadBucket
+	root   *cachedItem
+}
+
+// AddTopLevelBucket recursively reads the passed top-level bucket and adds
+// all content below it to the cache.
+func (c *Cache) AddTopLevelBucket(key []byte) error {
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	var root *cachedItem
+
+	err := View(c.backend, func(tx RTx) error {
```

Comment:
Hmm, so we'll end up recursively reading the _entire_ bucket here when triggered by a single read. Certain sections of the DB like the payment DB have several layers of nesting. Would be curious to gauge the impact/overhead of this vs just caching individual key/values then are accessed frequently.

Reply:
No, the idea is (see `channeldb/db.go`) to `AddTopLevelBucket()` on start and then keep things in memory. Unfortunately for the graph to work this way we need the channel state too.